From ef3e7219aa6ac8b2ab4f67359d51a5be9c8dfcc2 Mon Sep 17 00:00:00 2001
From: Dan Liang <dan.liang@intel.com>
Date: Tue, 30 Apr 2013 14:26:16 +0800
Subject: [PORT FROM MAIN] libstagefright: adjust audio seek for mkv container

BZ: 57484

The original seeking method can only deal with the case
that the SimpleBlocks of both audio and video are packed
in the head of a cluster. The testing clip is quite
special, as the key frame for video is located at the end
of a cluster, at 10000ms, while the first SimpleBlock of
audio in this cluster starts at 5013ms. So video is stopped
untill audio catches up.

After modification, audio will advance to the nearest point
as video has seeked to.

Change-Id: Ie9c62df10f2f64004ffde9e36bf8a67608246fff
Orig-Change-Id: I5bdde9db5cfc278aacf47f509a949d3ec90722b1
Signed-off-by: Dan Liang <dan.liang@intel.com>
---
 .../libstagefright/matroska/MatroskaExtractor.cpp  |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/media/libstagefright/matroska/MatroskaExtractor.cpp b/media/libstagefright/matroska/MatroskaExtractor.cpp
index ca9d2da..535ad3c 100644
--- a/media/libstagefright/matroska/MatroskaExtractor.cpp
+++ b/media/libstagefright/matroska/MatroskaExtractor.cpp
@@ -406,8 +406,8 @@ void BlockIterator::seek(
 
         if (eos()) break;
 
-        if (isAudio || block()->IsKey()) {
-            // Accept the first key frame
+        if((!isAudio && block()->IsKey()) ||
+            (isAudio && (blockTimeUs() >= seekTimeUs))) {
             *actualFrameTimeUs = (block()->GetTime(mCluster) + 500LL) / 1000LL;
             ALOGV("Requested seek point: %lld actual: %lld",
                   seekTimeUs, actualFrameTimeUs);
-- 
1.7.4.1

